File: bindings/java/examples/SimpleServer.java
import com.mtls.*;

import java.nio.charset.StandardCharsets;

/**
 * Simple mTLS server example.
 *
 * Demonstrates creating an mTLS listener and handling incoming connections.
 */
public class SimpleServer {
    public static void main(String[] args) {
        if (args.length < 3) {
            System.err.println("Usage: java SimpleServer <ca-cert> <server-cert> <server-key> [bind-address]");
            System.err.println("Example: java SimpleServer ca.pem server.pem server.key 0.0.0.0:8443");
            System.exit(1);
        }

        String caCert = args[0];
        String serverCert = args[1];
        String serverKey = args[2];
        String bindAddress = args.length > 3 ? args[3] : "0.0.0.0:8443";

        try {
            // Print library version
            System.out.println("mTLS Library Version: " + Context.getVersion());

            // Create configuration
            Config config = new Config.Builder()
                    .caCertFile(caCert)
                    .certFile(serverCert, serverKey)
                    .requireClientCert(true)  // Require client certificates
                    .verifyHostname(false)
                    .build();

            System.out.println("Starting server on " + bindAddress + "...");

            // Create context and listener
            try (Context ctx = new Context(config);
                 Listener listener = ctx.listen(bindAddress)) {

                System.out.println("Server listening on " + listener.getAddress());
                System.out.println("Press Ctrl+C to stop the server");
                System.out.println();

                // Accept connections in a loop
                int connectionCount = 0;
                while (true) {
                    System.out.println("Waiting for connection...");

                    try (Connection conn = listener.accept()) {
                        connectionCount++;
                        System.out.println("\n[Connection #" + connectionCount + "] New client connected!");
                        System.out.println("  Remote address: " + conn.getRemoteAddress());
                        System.out.println("  Local address: " + conn.getLocalAddress());

                        // Get peer identity
                        PeerIdentity identity = conn.getPeerIdentity();
                        if (identity != null) {
                            System.out.println("  Client Identity:");
                            System.out.println("    Common Name: " + identity.getCommonName());
                            System.out.println("    SANs: " + identity.getSubjectAltNames());
                            if (identity.hasSpiffeId()) {
                                System.out.println("    SPIFFE ID: " + identity.getSpiffeId());
                            }
                            System.out.println("    Valid: " + identity.isValid());
                        }

                        // Handle the connection
                        handleConnection(conn);

                        System.out.println("[Connection #" + connectionCount + "] Client disconnected\n");

                    } catch (MtlsException e) {
                        System.err.println("Connection error: " + e.getMessage());
                        // Continue accepting new connections
                    }
                }
            }

        } catch (MtlsException e) {
            System.err.println("mTLS Error: " + e.getMessage());
            System.err.println("Error code: " + e.getErrorCode());
            System.err.println("Category: " + e.getCategory());
            e.printStackTrace();
            System.exit(1);
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
            System.exit(1);
        }
    }

    private static void handleConnection(Connection conn) throws MtlsException {
        // Read data from client
        System.out.println("  Waiting for data...");
        byte[] buffer = new byte[4096];
        int bytesRead = conn.read(buffer);

        if (bytesRead > 0) {
            String message = new String(buffer, 0, bytesRead, StandardCharsets.UTF_8);
            System.out.println("  Received (" + bytesRead + " bytes): " + message);

            // Echo back to client
            String response = "Echo: " + message;
            byte[] responseBytes = response.getBytes(StandardCharsets.UTF_8);
            int bytesSent = conn.write(responseBytes);
            System.out.println("  Sent (" + bytesSent + " bytes): " + response);
        } else {
            System.out.println("  No data received");
        }
    }
}

File: bindings/rust/mtls/examples/simple_server.rs
//! Simple mTLS server example.
//!
//! Usage:
//!   cargo run --example simple_server -- <bind:port> <ca.pem> <server.pem> <server.key>

use std::env;
use std::io::{Read, Write};
use std::time::Duration;

use mtls::{Config, Context};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 5 {
        eprintln!(
            "Usage: {} <bind:port> <ca.pem> <server.pem> <server.key>",
            args[0]
        );
        std::process::exit(1);
    }

    let bind_addr = &args[1];
    let ca_cert = &args[2];
    let server_cert = &args[3];
    let server_key = &args[4];

    // Print library version
    println!("mTLS library version: {}", mtls::version());

    // Build configuration
    let config = Config::builder()
        .ca_cert_file(ca_cert)
        .cert_file(server_cert, server_key)
        .require_client_cert(true)
        .read_timeout(Duration::from_secs(60))
        .write_timeout(Duration::from_secs(60))
        .build()?;

    println!("Configuration created successfully");

    // Create context
    let ctx = Context::new(&config)?;
    println!("Context created successfully");

    // Start listener
    println!("Listening on {}...", bind_addr);
    let listener = ctx.listen(bind_addr)?;

    println!("Server listening on {}", listener.addr());

    println!("Waiting for connections...");

    // Accept and handle connections
    for conn_result in listener.incoming() {
        match conn_result {
            Ok(mut conn) => {
                println!("New connection!");

                // Print connection info
                if let Some(addr) = conn.remote_addr() {
                    println!("  Remote address: {}", addr);
                }

                // Print peer identity
                if let Some(identity) = conn.peer_identity() {
                    println!("  Peer CN: {}", identity.common_name);
                    if !identity.sans.is_empty() {
                        println!("  Peer SANs: {:?}", identity.sans);
                    }
                }

                // Read message
                let mut buf = [0u8; 1024];
                match conn.read(&mut buf) {
                    Ok(n) if n > 0 => {
                        let msg = String::from_utf8_lossy(&buf[..n]);
                        println!("  Received: {}", msg);

                        // Echo back
                        let response = format!("Echo: {}", msg);
                        if let Err(e) = conn.write_all(response.as_bytes()) {
                            eprintln!("  Write error: {}", e);
                        } else {
                            println!("  Sent response");
                        }
                    }
                    Ok(_) => {
                        println!("  Client disconnected");
                    }
                    Err(e) => {
                        eprintln!("  Read error: {}", e);
                    }
                }

                // Close connection
                conn.close();
                println!("  Connection closed");
            }
            Err(e) => {
                eprintln!("Accept error: {}", e);
            }
        }
    }

    Ok(())
}
